host/sgx/exception_sim.c

```diff
@@ -0,0 +1,226 @@
+    ssa_gpr->r15 = (uint64_t)(context->uc_mcontext.gregs[REG_R15]);
+    ssa_gpr->rip = (uint64_t)(context->uc_mcontext.gregs[REG_RIP]);
+    ssa_gpr->rflags = (uint64_t)(context->uc_mcontext.gregs[REG_EFL]);
+
+    // This flag is checked by virtual dispacher.
+    ssa_gpr->exit_info.as_fields.valid = true;
+}
+
+static void _update_sgx_vector(ucontext_t* context, int sig_num)
+{
+    // Hardcode here must match g_vector_to_exception_code_mapping[] in
+    // enclave/core/sgx/exception.c
+    uint32_t sgx_vector = (uint32_t)-1;
+    switch (sig_num)
+    {
+        case SIGILL: // OE_EXCEPTION_ILLEGAL_INSTRUCTION
+            sgx_vector = 6;
+            break;
+        case SIGTRAP: // OE_EXCEPTION_BREAKPOIN
+            sgx_vector = 3;
+            break;
+        case SIGBUS: // OE_EXCEPTION_MISALIGNMENT
+            sgx_vector = 17;
+            break;
+        case SIGFPE: // OE_EXCEPTION_DIVIDE_BY_ZERO
+            sgx_vector = 0;
+            break;
+        case SIGSEGV: // OE_EXCEPTION_ACCESS_VIOLATION
+            sgx_vector = 13;
+            break;
+    }
+
+    sgx_tcs_t* tcs = (sgx_tcs_t*)(context->uc_mcontext.gregs[REG_RBX]);
+    sgx_ssa_gpr_t* ssa_gpr = _get_ssa_gpr(tcs);
+    ssa_gpr->exit_info.as_fields.vector = sgx_vector;
+}
+
+static void _update_context_from_ssa(ucontext_t* context)
+{
+    sgx_tcs_t* tcs = (sgx_tcs_t*)(context->uc_mcontext.gregs[REG_RBX]);
+    sgx_ssa_gpr_t* ssa_gpr = _get_ssa_gpr(tcs);
+
+    context->uc_mcontext.gregs[REG_RAX] = (greg_t)ssa_gpr->rax;
+    context->uc_mcontext.gregs[REG_RBX] = (greg_t)ssa_gpr->rbx;
+    context->uc_mcontext.gregs[REG_RCX] = (greg_t)ssa_gpr->rcx;
+    context->uc_mcontext.gregs[REG_RDX] = (greg_t)ssa_gpr->rdx;
+    context->uc_mcontext.gregs[REG_RSP] = (greg_t)ssa_gpr->rsp;
+    context->uc_mcontext.gregs[REG_RBP] = (greg_t)ssa_gpr->rbp;
+    context->uc_mcontext.gregs[REG_RSI] = (greg_t)ssa_gpr->rsi;
+    context->uc_mcontext.gregs[REG_RDI] = (greg_t)ssa_gpr->rdi;
+    context->uc_mcontext.gregs[REG_R8] = (greg_t)ssa_gpr->r8;
+    context->uc_mcontext.gregs[REG_R9] = (greg_t)ssa_gpr->r9;
+    context->uc_mcontext.gregs[REG_R10] = (greg_t)ssa_gpr->r10;
+    context->uc_mcontext.gregs[REG_R11] = (greg_t)ssa_gpr->r11;
+    context->uc_mcontext.gregs[REG_R12] = (greg_t)ssa_gpr->r12;
+    context->uc_mcontext.gregs[REG_R13] = (greg_t)ssa_gpr->r13;
+    context->uc_mcontext.gregs[REG_R14] = (greg_t)ssa_gpr->r14;
+    context->uc_mcontext.gregs[REG_R15] = (greg_t)ssa_gpr->r15;
+    context->uc_mcontext.gregs[REG_RIP] = (greg_t)ssa_gpr->rip;
+    context->uc_mcontext.gregs[REG_EFL] = (greg_t)ssa_gpr->rflags;
+}
+
+static void _oe_eresume_sim(ucontext_t* context, void* enclave_fs)
+{
+    // Update cssa as ERESUME does in real mode.
+    sgx_tcs_t* tcs = (sgx_tcs_t*)(context->uc_mcontext.gregs[REG_RBX]);
+    tcs->cssa--;
+
+    // Change the FS register to enclave side.
+    oe_set_fs_register_base(enclave_fs);
```

Comment:
Would be good to save and restore both fs and gs.

Reply:
Good suggestion, now both fs/gs are processed, just like oe_enter_sim.